packages/nouns-contracts/contracts/client-incentives/Rewards.sol

```diff
@@ -0,0 +1,624 @@
+        t.rewardPerVote = t.votingRewardForPeriod / t.numEligibleVotes;
+
+        emit ProposalRewardsUpdated(
+            t.nextProposalIdToReward,
+            lastProposalId,
+            auctionRevenue,
+            t.rewardPerProposal,
+            t.rewardPerVote
+        );
+
+        //// Second loop over the proposals:
+        //// 1. Skip proposals that were deleted for non eligibility.
+        //// 2. Reward proposal's clientId.
+        //// 3. Reward the clientIds that faciliated voting.
+        //// 4. Make sure all voting clientIds were included.
+
+        InMemoryMapping.Mapping memory m = InMemoryMapping.createMapping({ maxClientId: nextTokenId() - 1 });
+
+        for (uint256 i; i < proposals.length; ++i) {
+            // skip non eligible deleted proposals
+            if (proposals[i].endBlock == 0) continue;
+
+            uint32 clientId = proposals[i].clientId;
+            if (clientId != 0) {
+                m.inc(clientId, t.rewardPerProposal);
+            }
+
+            uint256 votesInProposal;
+            NounsDAOTypes.ClientVoteData[] memory voteData = proposals[i].voteData;
+            for (uint256 j; j < votingClientIds.length; ++j) {
+                clientId = votingClientIds[j];
+                uint256 votes = voteData[j].votes;
+                require(votes > 0, 'all clientId must have votes');
+                if (clientId != 0) {
+                    m.inc(clientId, votes * t.rewardPerVote);
+                }
+                votesInProposal += votes;
+            }
+            require(
+                votesInProposal == proposals[i].forVotes + proposals[i].againstVotes + proposals[i].abstainVotes,
+                'not all votes accounted'
+            );
+        }
+
+        uint256 numValues = m.numValues();
+        for (uint32 i = 0; i < numValues; ++i) {
+            InMemoryMapping.ClientBalance memory cb = m.getValue(i);
+            $._clientBalances[cb.clientId] += cb.balance;
+            emit ClientRewarded(cb.clientId, cb.balance);
+        }
+
+        GasRefund.refundGas($.ethToken, startGas);
+    }
+
+    /**
+     * @notice Withdraws the balance of a client
+     * @dev The caller must be the owner of the NFT with id `clientId`
+     * @dev The maximum value of `amount` is one wei less than in `_clientBalances[clientId]`.
+     * This in order to leave 1 wei in storage and avoid expensive gas writes in future balance increases.
+     * @param clientId Which client balance to withdraw
+     * @param to the address to withdraw to
+     * @param amount amount of withdraw
+     */
+    function withdrawClientBalance(uint32 clientId, address to, uint256 amount) public whenNotPaused {
+        RewardsStorage storage $ = _getRewardsStorage();
+
+        require(ownerOf(clientId) == msg.sender, 'must be client NFT owner');
+        require(amount < $._clientBalances[clientId], 'amount too large');
+
+        $._clientBalances[clientId] -= amount;
```

Comment:
@davidbrai I think we're reading the client's balance twice in this function and could instead cache it in a local var, wdyt?

Reply:
@eladmallel yes it can.
not touching it for now because maybe you'll change the way this works?